LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs

```diff
@@ -30,96 +36,103 @@ internal sealed class ConcentratorDeduplication<T> :
-            var head = buffer; // keeps a view pointing at the start of the buffer
-            buffer = updf.DevAddr.Write(buffer);
-            buffer = updf.Mic.Write(buffer);
-            _ = Encoding.UTF8.GetBytes(updf.Payload, buffer);
-            BinaryPrimitives.WriteUInt16LittleEndian(buffer[updf.Payload.Length..], updf.Counter);
 
-            var key = Sha256.ComputeHash(head.ToArray());
+            var index = 0;
+            BinaryPrimitives.WriteUInt32LittleEndian(buffer, BinaryPrimitives.ReadUInt32LittleEndian(payload.DevAddr.Span));
+            index += payload.DevAddr.Length;
+
+            if (!payload.Mic.IsEmpty)
+                BinaryPrimitives.WriteUInt32LittleEndian(buffer[index..], BinaryPrimitives.ReadUInt32LittleEndian(payload.Mic.Span));
+            index += payload.Mic.Length;
+
+            payload.RawMessage?.CopyTo(buffer[index..]);
+            index += payload.RawMessage?.Length ?? 0;
+
+            BinaryPrimitives.WriteUInt16LittleEndian(buffer[index..], BinaryPrimitives.ReadUInt16LittleEndian(payload.Fcnt.Span));
+
+            var key = Sha256.ComputeHash(buffer.ToArray());
 
             return BitConverter.ToString(key);
         }
 
-        private static string CreateCacheKeyCore(JoinRequestFrame joinReq)
+        private static string CreateCacheKeyCore(LoRaPayloadJoinRequest payload)
         {
+            var joinEui = JoinEui.Read(payload.AppEUI.Span);
+            var devEui = DevEui.Read(payload.DevEUI.Span);
+            var devNonce = DevNonce.Read(payload.DevNonce.Span);
+
             var totalBufferLength = JoinEui.Size + DevEui.Size + DevNonce.Size;
             Span<byte> buffer = stackalloc byte[totalBufferLength];
             var head = buffer; // keeps a view pointing at the start of the buffer
-            buffer = joinReq.JoinEui.Write(buffer);
-            buffer = joinReq.DevEui.Write(buffer);
-#pragma warning disable IDE0058
-            // assigning to a discard results in compilation error CS8347
-            joinReq.DevNonce.Write(buffer);
-#pragma warning restore
+
+            buffer = joinEui.Write(buffer);
+            buffer = devEui.Write(buffer);
+            buffer = devNonce.Write(buffer);
 
             var key = Sha256.ComputeHash(head.ToArray());
 
             return BitConverter.ToString(key);
         }
 
-        private void AddToCache(string key, StationEui stationEui)
-            => this.cache.Set(key, stationEui, new MemoryCacheEntryOptions()
-            {
-                SlidingExpiration = DefaultExpiration
-            });
-
-        public void Dispose() => this.cache.Dispose();
+        private bool ShouldDrop(LoRaRequest loRaRequest, LoRaDevice? loRaDevice)
+            => loRaRequest.Payload is LoRaPayloadJoinRequest
+            || (loRaRequest.Payload is LoRaPayloadData && this.deduplicationStrategy.Create(loRaDevice) is DeduplicationStrategyDrop);
```

Comment:
I think we can simply look up the strategy on the LoRaDevice.Deduplication and avoid the creation of a object on the heap, that is garbage immediately.

Reply:
True, changed in bb8d581975b886a8e2ed87aca879d1e826b43d05